lib/auth/server_monitor.go

```diff
@@ -0,0 +1,148 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	insecurerand "math/rand"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	headerv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
+	notificationsv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/notifications/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/inventory"
+	"github.com/gravitational/teleport/lib/utils/interval"
+)
+
+const (
+	// systemClockCheckCycle is the period when system clock comparison is launched
+	// across all inventories, to be gathered for global notifications, if any.
+	systemClockCheckCycle = 10 * time.Minute
+	// systemClockThreshold is the duration threshold for triggering a warning
+	// if the time difference exceeds this threshold.
+	systemClockThreshold = time.Minute
+	// systemClockNotificationWarningName is the ID for adding the global notification.
+	systemClockNotificationWarningName = "cluster-monitor-system-clock-warning"
+)
+
+// MonitorSystemTime runs the periodic check for iterating through all inventories
+// to ping them and receive the system clock difference.
+func (a *Server) MonitorSystemTime(ctx context.Context) error {
+	checkInterval := interval.New(interval.Config{
+		FirstDuration: time.Minute,
+		Duration:      systemClockCheckCycle,
+		Clock:         a.GetClock(),
+		Jitter:        retryutils.NewSeventhJitter(),
+	})
+	defer checkInterval.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-checkInterval.Next():
+			var inventories []string
+			a.inventory.Iter(func(handle inventory.UpstreamHandle) {
```

Comment:
I don't think you're ever intended to use `(*inventory.Controller).Iter` with any visitor function that can block - it locks each shard of the inventory in order, and any new connection to the cluster will have to wait until the whole shard has been iterated through.